### ISSUE TITLE
Handle additional AVCs

### ISSUE THREAD

**Issue by tiran** [latchset/ipa-custodia-selinux#1]
@wrabcak here are some more AVCs with explanation.

## ipa-custodia start

```
type=AVC msg=audit(1492697537.948:2819): avc:  denied  { open } for  pid=17847 comm="ipa-custodia" path="/var/log/ipa-custodia.audit.log" dev="dm-0" ino=275886 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:var_log_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492697538.114:2820): avc:  denied  { execute } for  pid=17848 comm="ipa-custodia" name="ldconfig" dev="dm-0" ino=4793 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:ldconfig_exec_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492697538.114:2821): avc:  denied  { read open } for  pid=17848 comm="ipa-custodia" path="/usr/sbin/ldconfig" dev="dm-0" ino=4793 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:ldconfig_exec_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492697538.114:2822): avc:  denied  { execute_no_trans } for  pid=17848 comm="ipa-custodia" path="/usr/sbin/ldconfig" dev="dm-0" ino=4793 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:ldconfig_exec_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492697537.618:2818): avc:  denied  { write } for  pid=17847 comm="ipa-custodia" name="nss" dev="dm-0" ino=270508 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:sssd_var_lib_t:s0 tclass=sock_file permissive=1
type=AVC msg=audit(1492697538.153:2823): avc:  denied  { net_admin } for  pid=17847 comm="ipa-custodia" capability=12  scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=capability permissive=1
```

* /var/log/ipa-custodia.audit.log
* ldconfig is probably coming from Python's ctypes module. ctypes.util.find_library() runs ldconfig to locate shared library

## ipa-replica-install --setup-ca

```
type=AVC msg=audit(1492700622.817:3544): avc:  denied  { connectto } for  pid=19970 comm="httpd" path="/run/httpd/ipa-custodia.sock" scontext=system_u:system_r:httpd_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=unix_stream_socket permissive=1
type=AVC msg=audit(1492700622.829:3545): avc:  denied  { search } for  pid=20363 comm="ipa-custodia" name="19970" dev="proc" ino=264879 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=dir permissive=1
type=AVC msg=audit(1492700622.830:3546): avc:  denied  { read } for  pid=20363 comm="ipa-custodia" name="loginuid" dev="proc" ino=264880 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492700622.830:3547): avc:  denied  { open } for  pid=20363 comm="ipa-custodia" path="/proc/19970/loginuid" dev="proc" ino=264880 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492700622.830:3548): avc:  denied  { getattr } for  pid=20363 comm="ipa-custodia" path="/proc/19970/loginuid" dev="proc" ino=264880 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492700622.944:3549): avc:  denied  { dac_override } for  pid=20363 comm="ipa-custodia" capability=1  scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=capability permissive=1
type=AVC msg=audit(1492700622.945:3550): avc:  denied  { read } for  pid=20363 comm="ipa-custodia" name="password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492700622.945:3551): avc:  denied  { open } for  pid=20363 comm="ipa-custodia" path="/etc/pki/pki-tomcat/password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492700622.945:3552): avc:  denied  { getattr } for  pid=20363 comm="ipa-custodia" path="/etc/pki/pki-tomcat/password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492700841.361:3598): avc:  denied  { connectto } for  pid=19974 comm="httpd" path="/run/httpd/ipa-custodia.sock" scontext=system_u:system_r:httpd_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=unix_stream_socket permissive=1
```

* Custodia looks into /proc/PID and uses /proc/PID/loginuid to authenticate processes
* /etc/pki/pki-tomcat/password.conf is read to open the NSS database with private keys

## ipa-kra-install

```
type=AVC msg=audit(1492701017.714:3633): avc:  denied  { search } for  pid=20507 comm="ipa-custodia" name="20462" dev="proc" ino=275475 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=dir permissive=1
type=AVC msg=audit(1492701317.490:3734): avc:  denied  { connectto } for  pid=21947 comm="httpd" path="/run/httpd/ipa-custodia.sock" scontext=system_u:system_r:httpd_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=unix_stream_socket permissive=1
type=AVC msg=audit(1492701317.502:3735): avc:  denied  { search } for  pid=22045 comm="ipa-custodia" name="21947" dev="proc" ino=295678 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=dir permissive=1
type=AVC msg=audit(1492701317.502:3736): avc:  denied  { read } for  pid=22045 comm="ipa-custodia" name="loginuid" dev="proc" ino=298094 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701317.502:3737): avc:  denied  { open } for  pid=22045 comm="ipa-custodia" path="/proc/21947/loginuid" dev="proc" ino=298094 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701317.502:3738): avc:  denied  { getattr } for  pid=22045 comm="ipa-custodia" path="/proc/21947/loginuid" dev="proc" ino=298094 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:httpd_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701317.641:3739): avc:  denied  { dac_override } for  pid=22045 comm="ipa-custodia" capability=1  scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=capability permissive=1
type=AVC msg=audit(1492701317.641:3740): avc:  denied  { read } for  pid=22045 comm="ipa-custodia" name="password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701317.641:3741): avc:  denied  { open } for  pid=22045 comm="ipa-custodia" path="/etc/pki/pki-tomcat/password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701317.641:3742): avc:  denied  { getattr } for  pid=22045 comm="ipa-custodia" path="/etc/pki/pki-tomcat/password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
```

## ipa ca-add

```
type=AVC msg=audit(1492701579.053:3796): avc:  denied  { read } for  pid=22190 comm="ipa-custodia" name="password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.053:3795): avc:  denied  { dac_override } for  pid=22190 comm="ipa-custodia" capability=1  scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=capability permissive=1
type=AVC msg=audit(1492701579.054:3797): avc:  denied  { open } for  pid=22190 comm="ipa-custodia" path="/etc/pki/pki-tomcat/password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.054:3798): avc:  denied  { getattr } for  pid=22190 comm="ipa-custodia" path="/etc/pki/pki-tomcat/password.conf" dev="dm-0" ino=281608 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:pki_tomcat_etc_rw_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.085:3799): avc:  denied  { read } for  pid=22191 comm="pki" name="passwd" dev="dm-0" ino=281800 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:sssd_public_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.086:3800): avc:  denied  { open } for  pid=22191 comm="pki" path="/var/lib/sss/mc/passwd" dev="dm-0" ino=281800 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:sssd_public_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.086:3801): avc:  denied  { getattr } for  pid=22191 comm="pki" path="/var/lib/sss/mc/passwd" dev="dm-0" ino=281800 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:sssd_public_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.086:3802): avc:  denied  { write } for  pid=22191 comm="pki" name="nss" dev="dm-0" ino=270508 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:sssd_var_lib_t:s0 tclass=sock_file permissive=1
type=AVC msg=audit(1492701579.086:3803): avc:  denied  { connectto } for  pid=22191 comm="pki" path="/var/lib/sss/pipes/nss" scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:sssd_t:s0 tclass=unix_stream_socket permissive=1
type=AVC msg=audit(1492701579.133:3804): avc:  denied  { create } for  pid=22191 comm="pki" scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=tcp_socket permissive=1
type=AVC msg=audit(1492701579.133:3805): avc:  denied  { bind } for  pid=22191 comm="pki" scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=tcp_socket permissive=1
type=AVC msg=audit(1492701579.133:3806): avc:  denied  { node_bind } for  pid=22191 comm="pki" saddr=::1 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:node_t:s0 tclass=tcp_socket permissive=1
type=AVC msg=audit(1492701579.143:3807): avc:  denied  { execute } for  pid=22192 comm="pki" name="ldconfig" dev="dm-0" ino=4793 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:ldconfig_exec_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.143:3808): avc:  denied  { read open } for  pid=22192 comm="pki" path="/usr/sbin/ldconfig" dev="dm-0" ino=4793 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:ldconfig_exec_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.143:3809): avc:  denied  { execute_no_trans } for  pid=22192 comm="pki" path="/usr/sbin/ldconfig" dev="dm-0" ino=4793 scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:object_r:ldconfig_exec_t:s0 tclass=file permissive=1
type=AVC msg=audit(1492701579.913:3812): avc:  denied  { execmem } for  pid=22198 comm="java" scontext=system_u:system_r:ipa_custodia_t:s0 tcontext=system_u:system_r:ipa_custodia_t:s0 tclass=process permissive=1
```

* The FreeIPA custodia plugin run /usr/bin/pki to extract private key material for sub CAs. The command is a Python script that runs a Java command to interact with Dogtag PKI.

**Comment by wrabcak** [comment]
@tiran 
In my PR there are all AVCs fixed except: 
allow ipa_custodia_t self:capability dac_override;  <- It looks that permissions on testes system are too tight. Check: http://danwalsh.livejournal.com/69478.html

allow ipa_custodia_t self:process execmem;  <- This makes executable anon/private file mapping  that is writable, but this is caused by java.

**Comment by tiran** [comment]
I have opened https://pagure.io/freeipa/issue/6888 for ```DAC_OVERRIDE```.

AFAIK Java requires executable / writeable mmap for dynamic code generation, probably in Java's JIT. There is not much we can do about at the moment.

**Comment by tiran** [comment]
I talked to @frasertweedale. We have to stick to Java and execmem for now. For now there is no other command to extract the key material with asymmetric key wrapping. pk12util can only perform symmetric encryption.

**Comment by tiran** [comment]
Fixed by PR #2 / 93683c10fb53cfc839e7fbb4b64712eba1e65fe2
